### PR TITLE
Treat . as a nested field in field_map of text embedding processor

### PR DESCRIPTION
**Pull request by Sanjana679**
### Description
These changes will allow embeddings to be computed when using a nested source field for a text embedding processor as shown in the configuration below:  

```
PUT /_ingest/pipeline/nlp-ingest-pipeline
{
  "description": "An NLP ingest pipeline",
  "processors": [
    {
      "text_embedding": {
        "model_id": "",
        "field_map": {
          “message.text": “message_embedding”
        }
      }
    }
  ]
}

PUT /my-nlp-index
{
    "settings": {
        "index.knn": true,
        "default_pipeline": “nlp-ingest-pipeline”
    },
    "mappings": {
        "properties": {
            "message_embedding": {
                "type": "knn_vector",
                "dimension": 384,
                "method": {
                    "name": "hnsw",
                    "engine": "lucene"
                }
            },
            "message.text": { 
                "type": "text"            
            },
            "color": {
                "type": "text"
            }
        }
    }
}

PUT /my-nlp-index/_doc/1
{
  "message: {"text": "Blueberry smoothie"},
  "color": "blue"
}

GET /my-nlp-index/_doc/1
```

### Issues Resolved
[#110 ](https://github.com/opensearch-project/neural-search/issues/110)

### Check List
- [x] New functionality includes testing.
    - [x] All tests pass
- [ ] New functionality has been documented.
    - [ ] New functionality has javadoc added
- [x] Commits are signed as per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by martin-gaievski** [comment]
how many nested levels are we going to support? We need to state that if it's more than 2 levels, e.g.
`my_object.some_property_group.some_field`. Currently with mapping we can do multiple levels. @navneet1v what do you think?

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java [hunk 1]
```diff
@@ -154,6 +154,16 @@ Map<String, Object> buildMapWithProcessorKeyAndOriginalValue(IngestDocument inge
         for (Map.Entry<String, Object> fieldMapEntry : fieldMap.entrySet()) {
             String originalKey = fieldMapEntry.getKey();
             Object targetKey = fieldMapEntry.getValue();
+
+            int nestedDotIndex = originalKey.indexOf('.');
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:158]: This is a user-provided info, can we add basic validation if it's not already done as part of the processor/pipeline definition.

if multiple levels of nested fields are needed this code may need a rework

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java [hunk 2]
```diff
@@ -154,6 +154,16 @@ Map<String, Object> buildMapWithProcessorKeyAndOriginalValue(IngestDocument inge
...
+            if (nestedDotIndex != -1) {
+                Map<String, Object> temp = new LinkedHashMap<>();
```
Comment (martin-gaievski) [inline src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:160]: why do we need this map object? Can you please use more meaningful name for map variable

Comment (Sanjana679) [reply src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java]: I added a more meaningful name for the map variable.

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java (hunk 1 above)
Comment (Sanjana679) [reply src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:165]: There is some basic validation done in `validateEmbeddingConfiguration` which is run on `fieldMap` in the constructor in this file. However, I will add some extra validation for nested fields.

**Comment by navneet1v** [comment]
@Sanjana679 thanks for raising the PR. Can you please add output for these few cases too and also cover them in integration tests if not done already.

```
PUT /_ingest/pipeline/nlp-ingest-pipeline
{
  "description": "An NLP ingest pipeline",
  "processors": [
    {
      "text_embedding": {
        "model_id": "",
        "field_map": {
          “message.text": “message.message_embedding”
        }
      }
    }
  ]
}
```

**Comment by navneet1v** [comment]
@Sanjana679 I can within the diff there are lines which are coming because of the different formatter you might have in your local. Please fix that too and make sure that the lines which are actually modified are coming in the diff.

File: src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java [hunk 3]
```diff
@@ -154,9 +164,20 @@ Map<String, Object> buildMapWithProcessorKeyAndOriginalValue(IngestDocument inge
...
             Object targetKey = fieldMapEntry.getValue();
+
+            int nestedDotIndex = originalKey.indexOf('.');
+            if (nestedDotIndex != -1) {
+                Map<String, Object> newTargetKey = new LinkedHashMap<>();
+                newTargetKey.put(originalKey.substring(nestedDotIndex + 1), targetKey);
+                targetKey = newTargetKey;
+
+                originalKey = originalKey.substring(0, nestedDotIndex);
+            }
```
Comment (navneet1v) [inline src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:175]: @Sanjana679 can you please provide details how we are handling multiple level of nesting here?

Comment (Sanjana679) [reply src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java:175]: At the moment, I'm not currently handling multiple levels of nesting, as I initially thought it was only for one level. However, I will work on handling multiple levels of nesting.

**Comment by navneet1v** [comment]
@Sanjana679 can you also see with your changes this issue will be resolved or not: https://forum.opensearch.org/t/need-neural-search-plugin-to-support-nested-field-type-array-of-objects/16760 ?

**Comment by Sanjana679** [comment]
> @Sanjana679 can you also see with your changes this issue will be resolved or not: https://forum.opensearch.org/t/need-neural-search-plugin-to-support-nested-field-type-array-of-objects/16760 ?

At the moment, my changes don't resolve this issue but if I can easily resolve this issue I will work on it.

**Comment by navneet1v** [comment]
> > @Sanjana679 can you also see with your changes this issue will be resolved or not: https://forum.opensearch.org/t/need-neural-search-plugin-to-support-nested-field-type-array-of-objects/16760 ?
> 
> At the moment, my changes don't resolve this issue but if I can easily resolve this issue I will work on it.

I checked the question the way user is using is not correct.

can we check if the processor is created like this below:

```
{
	“description”: “MLsearchtestpipeline”,
	“processors”: [
		{
			“text_embedding”: {
				“model_id”: “8-S0XYsBFQTip4T18k-U”,
				“field_map”: {
					“question_embeddings.text”: "question_embeddings.text_embedding"
				}
			}
		}
	]
}
```
 will it work or not.

**Review by samuel-oci — COMMENTED** [review]
Can you please add tests to cover the new logic described here for nested fields ingestion?

**Comment by martin-gaievski** [comment]
implemented same logic in https://github.com/opensearch-project/neural-search/pull/841, we can close this PR

**Pull request closed without merging**